src/analysis.ts

```diff
@@ -61,3 +109,138 @@ export class Analysis {
     }
   }
 }
+
+
+class PackageGatherer implements AnalysisVisitor {
+  packageDirs = new Set<string>();
+  visitDocumentDescriptor(dd: DocumentDescriptor): void {
+    if (dd.document instanceof JsonDocument &&
+        (dd.document.url.endsWith('package.json') ||
+         dd.document.url.endsWith('bower.json'))) {
+      const dirname = path.dirname(dd.document.url);
+      if (!this.packageDirs.has(dirname)) {
+        this.packageDirs.add(trimLeft(dirname, '/'));
+      }
+    }
+  }
+}
+
+class ElementGatherer implements AnalysisVisitor {
+  elementDescriptors: ElementDescriptor[] = [];
+  locationOffsets = new Map<ElementDescriptor, LocationOffset>();
+  elementPaths = new Map<ElementDescriptor, string>();
+  visitElement(elementDescriptor: ElementDescriptor, path: Descriptor[]): void {
```

Comment:
`path` is shadowing the module here. Consider renaming the module to something like `pathLib`